File: agent_flow_graph.py
"""Normalize, validate, and order agent flow graphs (nodes + edges)."""

from __future__ import annotations

from collections import deque
from typing import Any


def empty_graph() -> dict[str, Any]:
    return {"v": 2, "nodes": [], "edges": []}


def node_kind(node: dict[str, Any] | None) -> str:
    if not node:
        return "agent"
    kind = str(node.get("kind") or "").strip().lower()
    if kind in {"task", "custom"}:
        return "task"
    if node.get("agent_id"):
        return "agent"
    if (node.get("instructions") or "").strip() or (node.get("title") or "").strip():
        return "task"
    return "agent"


def node_label(node: dict[str, Any] | None, *, fallback: str = "Step") -> str:
    if not node:
        return fallback
    if node_kind(node) == "task":
        return (node.get("title") or "").strip() or "Custom step"
    return (
        (node.get("agent_name") or "").strip()
        or (node.get("agent_slug") or "").strip()
        or fallback
    )


def _normalized_node(node: dict[str, Any], *, index: int) -> dict[str, Any] | None:
    node_id = str(node.get("id") or "").strip()
    if not node_id:
        return None
    kind = node_kind(node)
    column = node.get("column")
    try:
        column_i = int(column) % 2 if column is not None else index % 2
    except (TypeError, ValueError):
        column_i = index % 2
    if kind == "task":
        return {
            "id": node_id,
            "kind": "task",
            "title": (node.get("title") or "").strip(),
            "instructions": (node.get("instructions") or "").strip(),
            "column": column_i,
        }
    agent_id = node.get("agent_id")
    if not agent_id:
        return None
    out: dict[str, Any] = {
        "id": node_id,
        "kind": "agent",
        "agent_id": agent_id,
        "column": column_i,
    }
    if node.get("agent_name"):
        out["agent_name"] = node["agent_name"]
    if node.get("agent_slug"):
        out["agent_slug"] = node["agent_slug"]
    return out


def _is_keepable_node(node: Any) -> bool:
    if not isinstance(node, dict) or not node.get("id"):
        return False
    kind = node_kind(node)
    if kind == "task":
        return True
    return bool(node.get("agent_id"))


def linear_steps_to_graph(steps: list[dict[str, Any]]) -> dict[str, Any]:
    nodes: list[dict[str, Any]] = []
    edges: list[dict[str, Any]] = []
    prev_id: str | None = None
    prev_handoff = ""
    for index, step in enumerate(steps):
        if not isinstance(step, dict):
            continue
        node_id = str(step.get("id") or f"n{index}")
        raw = {**step, "id": node_id}
        normalized = _normalized_node(raw, index=index)
        if not normalized:
            continue
        nodes.append(normalized)
        if prev_id is not None:
            edges.append(
                {
                    "from": prev_id,
                    "to": node_id,
                    "handoff": prev_handoff,
                }
            )
        prev_handoff = (step.get("handoff") or "").strip()
        prev_id = node_id
    return {"v": 2, "nodes": nodes, "edges": edges}


def normalize_flow_steps(steps: Any) -> dict[str, Any]:
    if isinstance(steps, dict) and steps.get("v") == 2:
        raw_nodes = [n for n in (steps.get("nodes") or []) if _is_keepable_node(n)]
        nodes = []
        for index, node in enumerate(raw_nodes):
            normalized = _normalized_node(node, index=index)
            if normalized:
                nodes.append(normalized)
        edges = [
            e
            for e in (steps.get("edges") or [])
            if isinstance(e, dict) and e.get("from") and e.get("to")
        ]
        return {"v": 2, "nodes": nodes, "edges": edges}
    if isinstance(steps, list):
        return linear_steps_to_graph(steps)
    return empty_graph()


def validate_graph(graph: dict[str, Any]) -> tuple[bool, str]:
    nodes = graph.get("nodes") or []
    edges = graph.get("edges") or []
    if not nodes:
        return True, ""
    node_ids = {n["id"] for n in nodes}
    for edge in edges:
        if edge.get("from") not in node_ids or edge.get("to") not in node_ids:
            return False, "Flow connection references a missing step"
        if edge.get("from") == edge.get("to"):
            return False, "A step cannot connect to itself"
    indegree = {node_id: 0 for node_id in node_ids}
    adjacency: dict[str, list[str]] = {node_id: [] for node_id in node_ids}
    for edge in edges:
        src = edge["from"]
        dst = edge["to"]
        adjacency[src].append(dst)
        indegree[dst] += 1
    queue = deque([node_id for node_id, degree in indegree.items() if degree == 0])
    visited = 0
    while queue:
        current = queue.popleft()
        visited += 1
        for nxt in adjacency[current]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                queue.append(nxt)
    if visited != len(node_ids):
        return False, "Flow has a cycle — remove a connection and try again"
    return True, ""


def topological_order(graph: dict[str, Any]) -> list[str]:
    nodes = graph.get("nodes") or []
    edges = graph.get("edges") or []
    node_ids = [n["id"] for n in nodes]
    indegree = {node_id: 0 for node_id in node_ids}
    adjacency: dict[str, list[str]] = {node_id: [] for node_id in node_ids}
    for edge in edges:
        src = edge["from"]
        dst = edge["to"]
        if src not in adjacency or dst not in indegree:
            continue
        adjacency[src].append(dst)
        indegree[dst] += 1
    queue = deque([node_id for node_id in node_ids if indegree[node_id] == 0])
    order: list[str] = []
    while queue:
        current = queue.popleft()
        order.append(current)
        for nxt in adjacency[current]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                queue.append(nxt)
    if len(order) != len(node_ids):
        raise ValueError("Flow has a cycle")
    return order


def incoming_edges(graph: dict[str, Any], node_id: str) -> list[dict[str, Any]]:
    return [edge for edge in (graph.get("edges") or []) if edge.get("to") == node_id]


def node_by_id(graph: dict[str, Any], node_id: str) -> dict[str, Any] | None:
    for node in graph.get("nodes") or []:
        if node.get("id") == node_id:
            return node
    return None

File: web/src/utils/agentFlowGraph.ts
import type { AgentFlowGraph, AgentFlowGraphEdge, AgentFlowGraphNode } from "../types";

export function emptyAgentFlowGraph(): AgentFlowGraph {
  return { v: 2, nodes: [], edges: [] };
}

export function linearStepsToGraph(
  steps: Array<{
    agent_id?: string;
    handoff?: string;
    id?: string;
    kind?: "agent" | "task";
    title?: string;
    instructions?: string;
  }>,
): AgentFlowGraph {
  const nodes: AgentFlowGraphNode[] = [];
  const edges: AgentFlowGraphEdge[] = [];
  let prevId: string | null = null;
  let prevHandoff = "";

  steps.forEach((step, index) => {
    const kind = (step as { kind?: string }).kind === "task" || (!step.agent_id && ((step as { instructions?: string }).instructions || (step as { title?: string }).title))
      ? "task"
      : "agent";
    if (kind === "agent" && !step.agent_id) return;
    const nodeId = step.id ?? `n${index}`;
    nodes.push({
      id: nodeId,
      kind,
      agent_id: step.agent_id,
      column: (index % 2) as 0 | 1,
      agent_name: (step as { agent_name?: string }).agent_name,
      agent_slug: (step as { agent_slug?: string }).agent_slug,
      title: (step as { title?: string }).title,
      instructions: (step as { instructions?: string }).instructions,
    });
    if (prevId) {
      edges.push({
        from: prevId,
        to: nodeId,
        handoff: prevHandoff,
      });
    }
    prevHandoff = step.handoff ?? "";
    prevId = nodeId;
  });

  return { v: 2, nodes, edges };
}

export function parseAgentFlowSteps(steps: unknown): AgentFlowGraph {
  if (steps && typeof steps === "object" && !Array.isArray(steps) && (steps as AgentFlowGraph).v === 2) {
    const graph = steps as AgentFlowGraph;
    return {
      v: 2,
      nodes: (graph.nodes ?? []).filter(
        (node) => node.id && (node.kind === "task" || node.agent_id || (node.instructions || node.title)),
      ),
      edges: (graph.edges ?? []).filter((edge) => edge.from && edge.to),
    };
  }
  if (Array.isArray(steps)) {
    return linearStepsToGraph(steps);
  }
  return emptyAgentFlowGraph();
}

export function validateAgentFlowGraph(graph: AgentFlowGraph): string | null {
  const nodeIds = new Set(graph.nodes.map((node) => node.id));
  if (nodeIds.size === 0) return null;

  for (const edge of graph.edges) {
    if (!nodeIds.has(edge.from) || !nodeIds.has(edge.to)) {
      return "Flow connection references a missing step";
    }
    if (edge.from === edge.to) {
      return "A step cannot connect to itself";
    }
  }

  const indegree = new Map<string, number>();
  const adjacency = new Map<string, string[]>();
  for (const node of graph.nodes) {
    indegree.set(node.id, 0);
    adjacency.set(node.id, []);
  }
  for (const edge of graph.edges) {
    indegree.set(edge.to, (indegree.get(edge.to) ?? 0) + 1);
    adjacency.get(edge.from)?.push(edge.to);
  }

  const queue = [...indegree.entries()].filter(([, degree]) => degree === 0).map(([id]) => id);
  let visited = 0;
  while (queue.length > 0) {
    const current = queue.shift();
    if (!current) break;
    visited += 1;
    for (const next of adjacency.get(current) ?? []) {
      const degree = (indegree.get(next) ?? 0) - 1;
      indegree.set(next, degree);
      if (degree === 0) queue.push(next);
    }
  }

  if (visited !== graph.nodes.length) {
    return "Flow has a cycle — remove a connection and try again";
  }
  return null;
}

export function nodeKind(node: AgentFlowGraphNode): "agent" | "task" {
  if (node.kind === "task") return "task";
  if (node.agent_id) return "agent";
  if ((node.instructions || "").trim() || (node.title || "").trim()) return "task";
  return "agent";
}

export function nodeLabel(node: AgentFlowGraphNode, index: number): string {
  if (nodeKind(node) === "task") {
    return (node.title || "").trim() || `Custom step ${index + 1}`;
  }
  return node.agent_name ?? node.agent_slug ?? `Step ${index + 1}`;
}

export function updateNode(
  graph: AgentFlowGraph,
  nodeId: string,
  patch: Partial<AgentFlowGraphNode>,
): AgentFlowGraph {
  return {
    ...graph,
    nodes: graph.nodes.map((node) => (node.id === nodeId ? { ...node, ...patch } : node)),
  };
}

export function appendNode(
  graph: AgentFlowGraph,
  node: AgentFlowGraphNode,
  options?: { linkFromLast?: boolean },
): AgentFlowGraph {
  const edges = [...graph.edges];
  if (options?.linkFromLast && graph.nodes.length > 0) {
    const last = graph.nodes[graph.nodes.length - 1];
    if (last && !edges.some((edge) => edge.from === last.id && edge.to === node.id)) {
      edges.push({ from: last.id, to: node.id, handoff: "" });
    }
  }
  return { v: 2, nodes: [...graph.nodes, node], edges };
}

export function addEdge(
  graph: AgentFlowGraph,
  from: string,
  to: string,
): { graph: AgentFlowGraph; error?: string } {
  if (from === to) {
    return { graph, error: "A step cannot connect to itself" };
  }
  if (graph.edges.some((edge) => edge.from === from && edge.to === to)) {
    return { graph };
  }
  const next: AgentFlowGraph = {
    ...graph,
    edges: [...graph.edges, { from, to, handoff: "" }],
  };
  const validationError = validateAgentFlowGraph(next);
  if (validationError) {
    return { graph, error: validationError };
  }
  return { graph: next };
}

export function removeEdge(graph: AgentFlowGraph, from: string, to: string): AgentFlowGraph {
  return {
    ...graph,
    edges: graph.edges.filter((edge) => !(edge.from === from && edge.to === to)),
  };
}

export function toggleEdge(
  graph: AgentFlowGraph,
  from: string,
  to: string,
): AgentFlowGraph {
  const exists = graph.edges.some((edge) => edge.from === from && edge.to === to);
  return {
    ...graph,
    edges: exists
      ? graph.edges.filter((edge) => !(edge.from === from && edge.to === to))
      : [...graph.edges, { from, to, handoff: "" }],
  };
}

export function updateEdgeHandoff(
  graph: AgentFlowGraph,
  from: string,
  to: string,
  handoff: string,
): AgentFlowGraph {
  return {
    ...graph,
    edges: graph.edges.map((edge) =>
      edge.from === from && edge.to === to ? { ...edge, handoff } : edge,
    ),
  };
}

export function removeNode(graph: AgentFlowGraph, nodeId: string): AgentFlowGraph {
  return {
    v: 2,
    nodes: graph.nodes.filter((node) => node.id !== nodeId),
    edges: graph.edges.filter((edge) => edge.from !== nodeId && edge.to !== nodeId),
  };
}

export type RepositionPlacement = "before" | "after" | "append";

/** Move a node to another column and/or reorder it relative to siblings. */
export function repositionNode(
  graph: AgentFlowGraph,
  nodeId: string,
  options: { column: 0 | 1; targetId?: string | null; placement?: RepositionPlacement },
): AgentFlowGraph {
  const moving = graph.nodes.find((node) => node.id === nodeId);
  if (!moving) return graph;

  const nodes = graph.nodes.filter((node) => node.id !== nodeId);
  const { targetId, placement = "append" } = options;
  let column = options.column;

  let insertIdx = nodes.length;

  if (targetId && placement !== "append") {
    const targetIdx = nodes.findIndex((node) => node.id === targetId);
    if (targetIdx !== -1) {
      column = (nodes[targetIdx].column ?? 0) as 0 | 1;
      insertIdx = placement === "before" ? targetIdx : targetIdx + 1;
    }
  } else {
    let lastInCol = -1;
    for (let i = nodes.length - 1; i >= 0; i -= 1) {
      if ((nodes[i].column ?? 0) === column) {
        lastInCol = i;
        break;
      }
    }
    if (lastInCol !== -1) {
      insertIdx = lastInCol + 1;
    } else if (column === 0) {
      const firstCol1 = nodes.findIndex((node) => (node.column ?? 0) === 1);
      insertIdx = firstCol1 === -1 ? nodes.length : firstCol1;
    }
  }

  nodes.splice(insertIdx, 0, { ...moving, column });
  return { v: 2, nodes, edges: graph.edges };
}

export function incomingTargets(graph: AgentFlowGraph, nodeId: string): AgentFlowGraphEdge[] {
  return graph.edges.filter((edge) => edge.to === nodeId);
}

export function outgoingTargets(graph: AgentFlowGraph, nodeId: string): AgentFlowGraphEdge[] {
  return graph.edges.filter((edge) => edge.from === nodeId);
}
